Add an ExtractCopyright program to create/update debian/copyright

StephenMcConnel opened the pull request:
Also fix a couple of incorrect assembly acknowledgements of icu.net.dll

This new project specifies .Net 4.6.1 for what it's worth.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/libpalaso/514)
<!-- Reviewable:end -->

ermshiperete commented:
This looks pretty good. I have a few comments and suggestions.

You'll see that I wondered about several methods if they have to be public. Later on when I came to the unit tests I realized that the methods are public for testing purposes. However, it might be cleaner to make them internal and expose the internal methods to the test project (by adding whatever attribute to the AssemblyInfo file). Having them with internal visibility makes it clear that they won't be called directly (except in unit tests).

---

Reviewed 14 of 14 files at r1.
Review status: all files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[ExtractCopyright/CopyrightFile.cs, line 38 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-Kjv_KvOr0Dwws_goZ3g:-Kjv_KvPbanpK5KlU7I6:biz32br) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L38)):*
> ```Smalltalk
> 		/// <param name="assemblyFolder">Assembly folder.</param>
> 		/// <param name="prefixFolder">Prefix folder.</param>
> 		static public int CreateOrUpdateCopyrightFile(string debianFolder, string prefixFolder)
> ```

I think the usual order is to have `public` first.

---

*[ExtractCopyright/CopyrightFile.cs, line 82 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvpJGBWOfeDyd8dONe:-KjvpJGBWOfeDyd8dONf:bxjhuxm) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L82)):*
> ```Smalltalk
> 		}
> 
> 		public static void ParseControlContentForValues(DebianControl controlFile, ref string programName, ref string contactEmail, ref string sourceUrl)
> ```

Does this have to be `public`? Then we should add a method comment.

---

*[ExtractCopyright/CopyrightFile.cs, line 100 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvoqxA5LfivIw-Qd5k:-KjvoqxCv1A_a6zRUN7m:b38ak39) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L100)):*
> ```Smalltalk
> 				if (maintainerField != null)
> 					contactEmail = maintainerField.Value;
> 				var urlField = sourcePara.FindField("Vcs-Browser");
> ```

This part on how to find the sourceUrl could either use a comment (e.g. `find sourceUrl from either Vcs-Browser, HomePage, or Vcs-Git`), or refactor to a separate method like `RetrieveSourceUrl`.

---

*[ExtractCopyright/CopyrightFile.cs, line 110 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvpNGXrh0uOwY5b8MT:-KjvpNGXrh0uOwY5b8MU:bxjhuxm) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L110)):*
> ```Smalltalk
> 		}
> 
> 		public static void ParseChangelogContentForValues(string[] lines, ref string programName, ref string contactEmail)
> ```

Does this have to be `public`? Then we should add a method comment.

---

*[ExtractCopyright/CopyrightFile.cs, line 116 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvphTpnhppQbPk-u_o:-KjvphTpnhppQbPk-u_p:bm8bivt) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L116)):*
> ```Smalltalk
> 				if (String.IsNullOrEmpty(programName) && lines[i].Contains(" urgency=") && lines[i].Contains(";"))
> 				{
> 					var headerPieces = lines[i].Trim().Split(new char[]{' '});
> ```

Might be good to add a comment with a sample line that we're trying to process.

---

*[ExtractCopyright/CopyrightFile.cs, line 121 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-Kjvpivt0fSFx9cNe4UT:-Kjvpivt0fSFx9cNe4UU:bm8bivt) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L121)):*
> ```Smalltalk
> 				else if (String.IsNullOrEmpty(contactEmail) && lines[i].StartsWith(" -- ") && lines[i].Contains("@"))
> 				{
> 					var line = lines[i].Substring(4).Trim();
> ```

Might be good to add a comment with a sample line that we're trying to process.

---

*[ExtractCopyright/CopyrightFile.cs, line 135 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvpmN8DnUTOQRd28ZW:-KjvpmN8DnUTOQRd28ZX:b-fttxh2) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L135)):*
> ```Smalltalk
> 		/// </summary>
> 		/// <returns>A new CopyrightFile object</returns>
> 		static public CopyrightFile CreateNewCopyrightFile(string programName, string contactEmail, string sourceUrl)
> ```

`public` usually comes first

---

*[ExtractCopyright/CopyrightFile.cs, line 160 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvqFhDryUiM5IlG6R5:-KjvqFhEMAWiyxZLWbni:bxjhuxm) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L160)):*
> ```Smalltalk
> 		}
> 
> 		public void AddOrUpdateParagraphFromAcknowledgement(Acknowledgements.AcknowledgementAttribute ack, string prefix)
> ```

Does this have to be `public`? Then we should add a method comment.

---

*[ExtractCopyright/CopyrightFile.cs, line 217 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvqerKxw48xk8XSq-h:-KjvqerLyruGK4XpKqnC:b9x5pja) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L217)):*
> ```Smalltalk
> 		}
> 
> 		private bool FilterForWindowsSpecific(string name)
> ```

A better method name would be something like `IsWindowsSpecificFile`

---

*[ExtractCopyright/CopyrightFile.cs, line 235 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-Kjvqt5eQLhRq83YTmJU:-Kjvqt5f4TJgbdJUQW86:bxjhuxm) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L235)):*
> ```Smalltalk
> 		}
> 
> 		public void AddLicenseParagraphIfNeeded(string license, IEnumerable<string> detailLines)
> ```

Does this have to be `public`? Then we should add a method comment.

---

*[ExtractCopyright/DebianControl.cs, line 46 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrWOfHIAIM-h_XzgT:-KjvrWOfHIAIM-h_XzgU:b2d33l2) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L46)):*
> ```Smalltalk
> 		}
> 
> 		public void ReadDebianControlFile(string filepath)
> ```

Does this have to be public?

---

*[ExtractCopyright/DebianControl.cs, line 57 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrTF00FpyUv0qpBr2:-KjvrTF00FpyUv0qpBr3:b2d33l2) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L57)):*
> ```Smalltalk
> 		}
> 
> 		public void ReadDebianParagraphs(TextReader reader)
> ```

Does this have to be public?

---

*[ExtractCopyright/DebianControl.cs, line 112 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrddRXDMg_fgNdsWN:-KjvrddSCIHSL7sMHUqY:b2d33l2) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L112)):*
> ```Smalltalk
> 		}
> 
> 		public void WriteDebianParagraphs(TextWriter writer)
> ```

Does this have to be public?

---

*[ExtractCopyright/DebianControl.cs, line 135 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrlGChxcxiJJmJ_cN:-KjvrlGChxcxiJJmJ_cO:b1kd213) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L135)):*
> ```Smalltalk
> 
> 
> 	public class DebianField
> ```

Make internal? Move to separate file?

---

*[ExtractCopyright/DebianControl.cs, line 167 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrpKaQtzKEaRInP95:-KjvrpKaQtzKEaRInP96:b-n15zxn) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L167)):*
> ```Smalltalk
> 	}
> 
> 	public class DebianParagraph
> ```

Make internal?

---

*[ExtractCopyright/Program.cs, line 48 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvsEXJyUo3Binpg5ap:-KjvsEXJyUo3Binpg5aq:ba2ofru) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/Program.cs#L48)):*
> ```Smalltalk
> 				while (!string.IsNullOrEmpty(current) && !Directory.Exists(debianFolder))
> 				{
> 					current = Path.GetDirectoryName(current);
> ```

Does this stop at the root folder? Or does Path.GetDirectoryName("/") return "/"?xn

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-:-KjvtIGgtPvpjJbCvzFS:b5x63ei)*
<!-- Sent from Reviewable.io -->

gmartin7 commented on ExtractCopyright/CopyrightFile.cs:
```diff
@@ -0,0 +1,375 @@
+			case "NAudio.dll":
+			case "KeymanLink.dll":
+			// part of Mono runtime, not part of a specific program
+			case "Mono.Posix":
```
So this isn't supposed to be Mono.Posix.dll?

gmartin7 commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[ExtractCopyright/CopyrightFile.cs, line 226 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjxYr5HvEnKpTO5FKdP:-KjxYr5HvEnKpTO5FKdQ:bsb2aqa) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L226)):*
> Mono.Posix
So this isn't supposed to be "Mono.Posix.dll"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/514)*
<!-- Sent from Reviewable.io -->

StephenMcConnel commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[ExtractCopyright/CopyrightFile.cs, line 38 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-Kjv_KvOr0Dwws_goZ3g:-KjyKu23Qwg4ktziuh8R:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L38)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

I think the usual order is to have `public` first.
</blockquote></details>

Done.

---

*[ExtractCopyright/CopyrightFile.cs, line 82 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvpJGBWOfeDyd8dONe:-KjyL6p5zdV4ItSB83wz:b-xonunc) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L82)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this have to be `public`? Then we should add a method comment.
</blockquote></details>

changed to internal, with tests enabled to use it.

---

*[ExtractCopyright/CopyrightFile.cs, line 100 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvoqxA5LfivIw-Qd5k:-KjyN9P_gHdNYSDeXEWe:b-p1i752) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L100)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

This part on how to find the sourceUrl could either use a comment (e.g. `find sourceUrl from either Vcs-Browser, HomePage, or Vcs-Git`), or refactor to a separate method like `RetrieveSourceUrl`.
</blockquote></details>

added comment

---

*[ExtractCopyright/CopyrightFile.cs, line 110 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvpNGXrh0uOwY5b8MT:-KjyLKLUeTGKbUKhHcsu:bfns1br) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L110)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this have to be `public`? Then we should add a method comment.
</blockquote></details>

changed to internal with tests enabled

---

*[ExtractCopyright/CopyrightFile.cs, line 116 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvphTpnhppQbPk-u_o:-KjyQi0bQwHX_GgKvUF8:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L116)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Might be good to add a comment with a sample line that we're trying to process.
</blockquote></details>

Done.

---

*[ExtractCopyright/CopyrightFile.cs, line 121 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-Kjvpivt0fSFx9cNe4UT:-KjyQihp2iX1p61QIi-0:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L121)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Might be good to add a comment with a sample line that we're trying to process.
</blockquote></details>

Done.

---

*[ExtractCopyright/CopyrightFile.cs, line 135 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvpmN8DnUTOQRd28ZW:-KjyLQn4me79U3W4qn5Q:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L135)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

`public` usually comes first
</blockquote></details>

Done.

---

*[ExtractCopyright/CopyrightFile.cs, line 160 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvqFhDryUiM5IlG6R5:-KjyLekLD3boLoPHinv7:bfns1br) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L160)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this have to be `public`? Then we should add a method comment.
</blockquote></details>

changed to internal with tests enabled

---

*[ExtractCopyright/CopyrightFile.cs, line 217 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvqerKxw48xk8XSq-h:-KjyO3zrqdU1RInQHMar:b-896fix) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L217)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

A better method name would be something like `IsWindowsSpecificFile`
</blockquote></details>

Done.

---

*[ExtractCopyright/CopyrightFile.cs, line 226 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjxYr5HvEnKpTO5FKdP:-KjyNeCeYHH7ANEIykIu:b-q11wg0) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L226)):*
<details><summary><i>Previously, gmartin7 (Gordon Martin) wrote…</i></summary><blockquote>

> Mono.Posix
So this isn't supposed to be "Mono.Posix.dll"?
</blockquote></details>

It's currently given as Mono.Posix, and we don't need it for the Debian copyright file because it's part of Mono, so I gave it a pass.  :-)

---

*[ExtractCopyright/CopyrightFile.cs, line 235 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-Kjvqt5eQLhRq83YTmJU:-KjyNVq6j9586WAyaZFc:bfns1br) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/CopyrightFile.cs#L235)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this have to be `public`? Then we should add a method comment.
</blockquote></details>

changed to internal with tests enabled

---

*[ExtractCopyright/DebianControl.cs, line 46 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrWOfHIAIM-h_XzgT:-KjyT1be57WRc03tuhyA:b9706y9) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L46)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this have to be public?
</blockquote></details>

changed to internal

---

*[ExtractCopyright/DebianControl.cs, line 57 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrTF00FpyUv0qpBr2:-KjyT42rmwa-t_uDslsF:b9706y9) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L57)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this have to be public?
</blockquote></details>

changed to internal

---

*[ExtractCopyright/DebianControl.cs, line 112 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrddRXDMg_fgNdsWN:-KjyT5nj03c7oLi5hSKD:b9706y9) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L112)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this have to be public?
</blockquote></details>

changed to internal

---

*[ExtractCopyright/DebianControl.cs, line 135 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrlGChxcxiJJmJ_cN:-KjyXJQ-Nl_U3aFRv5tR:b-oys2kf) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L135)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Make internal? Move to separate file?
</blockquote></details>

moved to a separate file and added a summary description

---

*[ExtractCopyright/DebianControl.cs, line 167 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvrpKaQtzKEaRInP95:-KjyXONiNpJNTyrDRfO4:b-oys2kf) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/DebianControl.cs#L167)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Make internal?
</blockquote></details>

moved to a separate file and added a summary description

---

*[ExtractCopyright/Program.cs, line 48 at r1](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-KjvsEXJyUo3Binpg5ap:-KjyTI8xSQgCD9A0QeJl:b-7kbydv) ([raw file](https://github.com/sillsdev/libpalaso/blob/47c5adec1126fe7ea0a74cbfaf74e148b7957275/ExtractCopyright/Program.cs#L48)):*
<details><summary><i>Previously, ermshiperete (Eberhard Beilharz) wrote…</i></summary><blockquote>

Does this stop at the root folder? Or does Path.GetDirectoryName("/") return "/"?xn
</blockquote></details>

Path.GetDirectoryName("/") returns null, hence the tests for IsNullOrEmpty,  I could have used == null instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/514)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/514)*
<!-- Sent from Reviewable.io -->

ermshiperete commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-:-Kk9-AT-QE8lrfCwKJIJ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

gmartin7 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/sillsdev/libpalaso/514#-:-KkBXlDZOYmS0o8WNMJ9:bnfp4nl)*
<!-- Sent from Reviewable.io -->
